ecdsa/src/rfc6979.rs

```diff
@@ -28,82 +29,47 @@ where
     Scalar<C>:
         PrimeField<Repr = FieldBytes<C>> + FromDigest<C> + Invert<Output = Scalar<C>> + Zeroize,
 {
-    let mut x = secret_scalar.to_repr();
-    let h1 = Scalar::<C>::from_digest(msg_digest).to_repr();
-    let mut hmac_drbg = HmacDrbg::<D>::new(&x, &h1, additional_data);
+    let bits_to_bytes = |b: elliptic_curve::ScalarBits<C>| {
+        // make byte buffer of the right size
+        let mut tmp = FieldBytes::<C>::default();
+        // convert bitslice from internal repr based on u32 or u64 to one based on u8
+        tmp.as_mut_slice()
+            .view_bits_mut::<Lsb0>()
+            .clone_from_bitslice(b.as_bitslice());
+        // convert to big endian
+        tmp.reverse();
+        tmp
+    };
+
+    // Verify that values roundtrip through bits_to_bytes and from_repr as expected.
+    // The from_repr interface does not guarantee big endian interpretation, and if it is not
+    // as expected the generated value may not have the required entropy.
+    let one = Scalar::<C>::one();
+    let one_bytes = bits_to_bytes(one.to_le_bits());
+    assert!(
+        one_bytes[one_bytes.len() - 1] == 1 && Scalar::<C>::from_repr(one_bytes) == Some(one),
+        "curve repr not as expected"
+    );
```

Comment:
This is obviously kind of ugly, but I don't see how I can avoid doing something like this with the current APIs, at least not without leaving a gaping hole in the API allowing for misuse. In theory at least the compiler should be able to fully evaluate this at compile time, but I kind of doubt it's that good.